### ISSUE TITLE
[18.0] res.currency.rate   error

### ISSUE THREAD

**Issue by yanghongeer** [odoo/odoo#190315]
### Odoo Version

- [ ] 16.0
- [ ] 17.0
- [X] 18.0
- [ ] Other (specify)

### Steps to Reproduce

1. open technical - Scheduled Actions - Currency: rate update
2.add   rate_usd = env['res.currency.rate'].search([('display_name', '=', 'USD')], limit=1)  , then run ,will show error .

![1](https://github.com/user-attachments/assets/1011cad0-e1c5-4090-b068-801441281e9f)
![2](https://github.com/user-attachments/assets/a276638d-1aed-44cb-9ba2-1054985a0e2c)

### Log Output

```shell
Odoo Server Error

Occured on on model ir.cron and id 16 on 2024-12-11 07:34:18 GMT

Traceback (most recent call last):
  File "/home/odoo18/odoo/odoo/tools/safe_eval.py", line 389, in safe_eval
    return unsafe_eval(c, globals_dict, locals_dict)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "ir.actions.server(300,)", line 3, in <module>
  File "/home/odoo18/odoo/odoo/models.py", line 1717, in search
    return self.search_fetch(domain, [], offset=offset, limit=limit, order=order)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/models.py", line 1741, in search_fetch
    query = self._search(domain, offset=offset, limit=limit, order=order or self._order)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/models.py", line 5736, in _search
    query = self._where_calc(domain)
            ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/models.py", line 5487, in _where_calc
    return expression.expression(domain, self).query
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/osv/expression.py", line 787, in __init__
    self.parse()
  File "/home/odoo18/odoo/odoo/osv/expression.py", line 1434, in parse
    push_result(model._condition_to_sql(alias, left, operator, right, self.query))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/models.py", line 3193, in _condition_to_sql
    sql_value = SQL("%s", field.convert_to_column(value, self, validate=False))
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/fields.py", line 1628, in convert_to_column
    value_float = value = float(value or 0.0)
                          ^^^^^^^^^^^^^^^^^^^
ValueError: could not convert string to float: 'USD'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/odoo18/odoo/odoo/http.py", line 1955, in _transactioning
    return service_model.retrying(func, env=self.env)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/service/model.py", line 137, in retrying
    result = func()
             ^^^^^^
  File "/home/odoo18/odoo/odoo/http.py", line 1922, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/http.py", line 2169, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/addons/base/models/ir_http.py", line 329, in _dispatch
    result = endpoint(**request.params)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/http.py", line 727, in route_wrapper
    result = endpoint(self, *args, **params_ok)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/addons/web/controllers/dataset.py", line 40, in call_button
    action = call_kw(request.env[model], method, args, kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/api.py", line 517, in call_kw
    result = getattr(recs, name)(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/addons/base/models/ir_cron.py", line 120, in method_direct_trigger
    self.ir_actions_server_id.run()
  File "/home/odoo18/odoo/odoo/addons/base/models/ir_actions.py", line 991, in run
    res = runner(run_self, eval_context=eval_context)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/addons/base/models/ir_actions.py", line 823, in _run_action_code_multi
    safe_eval(self.code.strip(), eval_context, mode="exec", nocopy=True, filename=str(self))  # nocopy allows to return 'action'
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo18/odoo/odoo/tools/safe_eval.py", line 403, in safe_eval
    raise ValueError('%r while evaluating\n%r' % (e, expr))
ValueError: ValueError("could not convert string to float: 'USD'") while evaluating
'model.run_update_currency()\n\nrate_usd = env[\'res.currency.rate\'].search([(\'display_name\', \'=\', \'USD\')], limit=1)\nrate_usd_line = rate_usd.search([(\'name\', \'!=\', "2000-1-1" )], order=\'create_date desc\', limit=1)'

The above server error caused the following client error:
RPC_ERROR: Odoo Server Error
    RPC_ERROR
        at makeErrorFromResponse (https://web/assets/e79b93a/web.assets_web.min.js:3092:163)
        at XMLHttpRequest.<anonymous> (https:///web/assets/e79b93a/web.assets_web.min.js:3097:13)
```

**Comment by intero-chz** [comment]
Searching on `display_name` will lead to a search on two fields in this case. Because the default model's `display_name` field is computed (if not redefined on the model, which is not the case for currency rates). In this default case odoo will search of the `_rec_name` definition of a model OR on the field names defined in `_rec_name_search` which is defined for currency rates (see [here](https://github.com/odoo/odoo/blob/f8007796406082a65023d959dc629ac153b619cf/odoo/addons/base/models/res_currency.py#L333)).

This `_rec_name_search` definition is really weird, because those two fields are no text fields at all. `name` is a Date field and `rate` a Float field. The second one is leading to the error.

TL;DR i don't recommend searching on `display_name` on model `res.currency.rate` because that's fucked up.

What are the alternatives:

1. search on the related currency's display name: `[("currency_id.display_name", "=", "USD")]`
2. just use the external ID of the [currency](https://github.com/odoo/odoo/blob/63a842c1a3201eefeda91d9a9b06eccff6c18bbd/odoo/addons/base/data/res_currency_data.xml#L5): `rate_usd = env.ref("base.USD")`

**Comment by yanghongeer** [comment]
OK，thank you very much
